### PR TITLE
dashboard part 1: add sync state and collect data for network dashboard

### PR DESCRIPTION
**Pull request by ping-ke**
issue: https://github.com/ethstorage/es-node/issues/253
To add grafana dashboard to overview the network states and this pr is adding sync state types and collecting state for network dashboard report later.

### CODE REVIEW

File: ethstorage/p2p/protocol/syncserver.go [hunk 1]
```diff
@@ -175,6 +182,7 @@ func (srv *SyncServer) handleGetBlobsByRangeRequest(ctx context.Context, stream
 		}
 	}
 	srv.metrics.ServerReadBlobs(peerID.String(), read, sucRead, time.Since(start))
+	srv.providedBlobs[req.ShardId] += uint64(len(res.Blobs))
```
Comment (qzhodl) [inline ethstorage/p2p/protocol/syncserver.go:185]: is there race condition?

File: ethstorage/p2p/protocol/syncserver.go [hunk 2]
```diff
@@ -79,9 +80,15 @@ func NewSyncServer(cfg *rollup.EsConfig, storageManager StorageManagerReader, m
 	if m == nil {
 		m = metrics.NoopMetrics
 	}
+
+	providedBlobs := make(map[uint64]uint64)
+	for _, shardId := range storageManager.Shards() {
+		providedBlobs[shardId] = 0
```
Comment (qzhodl) [inline ethstorage/p2p/protocol/syncserver.go:86]: Do we want to persist this data into the database? Why do we choose a different handling way with the mining status?

File: ethstorage/p2p/protocol/syncserver.go [hunk 3]
```diff
@@ -311,3 +320,7 @@ func (srv *SyncServer) HandleRequestShardList(ctx context.Context, log log.Logge
 	}
 	log.Debug("Write response done for HandleRequestShardList")
 }
+
+func (srv *SyncServer) ProvidedBlobs() map[uint64]uint64 {
+	return srv.providedBlobs
```
Comment (qzhodl) [inline ethstorage/p2p/protocol/syncserver.go:325]: This is a different handling way with the sync client, and also it is different with the mining status

File: ethstorage/p2p/protocol/syncclient.go [hunk 4]
```diff
@@ -234,41 +229,48 @@ func (s *SyncClient) setSyncDone() {
...
-			for _, task := range progress.Tasks {
-				log.Debug("Load sync subTask", "contract", task.Contract.Hex(),
-					"shard", task.ShardId, "count", len(task.SubTasks))
-				task.healTask = &healTask{
+			for _, t := range progress.Tasks {
+				// TODO if t.State is nil, that mean the status is marshal by old state,
+				// set process value to SyncState to make it compatible.
+				// it can be removed after public test done.
+				if t.State == nil {
+					t.State = &SyncState{
```
Comment (qzhodl) [inline ethstorage/p2p/protocol/syncclient.go:247]: what if a user is syncing, the blobsSynced will be reset to zero and the report process would still be correct?

File: ethstorage/p2p/protocol/syncserver.go [hunk 5]
```diff
@@ -311,3 +349,38 @@ func (srv *SyncServer) HandleRequestShardList(ctx context.Context, log log.Logge
...
+	}
+
+	err = srv.db.Put(providedBlobsKey, states)
+	if err != nil {
+		log.Error("Failed to store provided blobs states", "err", err)
+		return
+	}
+}
+
+func (srv *SyncServer) SaveProvidedBlobs() {
```
Comment (syntrust) [inline ethstorage/p2p/protocol/syncserver.go:369]: Who calls (srv *SyncServer) SaveProvidedBlobs()?

**Review by syntrust — APPROVED** [review]

**Review by qzhodl — APPROVED** [review]

**Pull request merged**